### ISSUE TITLE
setObjectOfFriendDecl(bool): Assertion "namespace includes other than ordinary or tag"' failed

### ISSUE THREAD

**Issue by balazske** [Ericsson/clang#318]
When analyzing a 'minilink' failed case with newer commits included the following assertion is triggered:
```
clang-5.0: clang/clang1/llvm/tools/clang/include/clang/AST/DeclBase.h:1041: void clang::Decl::setObjectOfFriendDecl(bool): Assertion `!(OldNS & ~(IDNS_Tag | IDNS_Ordinary | IDNS_Type | IDNS_TagFriend | IDNS_OrdinaryFriend | IDNS_LocalExtern)) && "namespace includes other than ordinary or tag"' failed.
#0 0x00007f822c639fd9 llvm::sys::PrintStackTrace(llvm::raw_ostream&) clang/clang1/llvm/lib/Support/Unix/Signals.inc:398:0
#1 0x00007f822c63a06a PrintStackTraceSignalHandler(void*) clang/clang1/llvm/lib/Support/Unix/Signals.inc:462:0
#2 0x00007f822c6384b0 llvm::sys::RunSignalHandlers() clang/clang1/llvm/lib/Support/Signals.cpp:49:0
#3 0x00007f822c639971 SignalHandler(int) clang/clang1/llvm/lib/Support/Unix/Signals.inc:252:0
#4 0x00007f8228f52cb0 (/lib/x86_64-linux-gnu/libc.so.6+0x36cb0)
#5 0x00007f8228f52c37 gsignal /build/eglibc-ripdx6/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56:0
#6 0x00007f8228f56028 abort /build/eglibc-ripdx6/eglibc-2.19/stdlib/abort.c:91:0
#7 0x00007f8228f4bbf6 __assert_fail_base /build/eglibc-ripdx6/eglibc-2.19/assert/assert.c:92:0
#8 0x00007f8228f4bca2 (/lib/x86_64-linux-gnu/libc.so.6+0x2fca2)
#9 0x00007f8224baf2a5 clang::Decl::setObjectOfFriendDecl(bool) clang/clang1/llvm/tools/clang/include/clang/AST/DeclBase.h:1043:0
#10 0x00007f8224b94ca6 clang::ASTNodeImporter::VisitFriendDecl(clang::FriendDecl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:2799:0
#11 0x00007f8224bbeb3d clang::declvisitor::Base<clang::declvisitor::make_ptr, clang::ASTNodeImporter, clang::Decl*>::Visit(clang::Decl*) clang/clang1/build/tools/clang/include/clang/AST/DeclNodes.inc:71:0
#12 0x00007f8224bab4a7 clang::ASTImporter::Import(clang::Decl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6835:0
#13 0x00007f8224b8b6b5 clang::ASTNodeImporter::ImportDeclContext(clang::DeclContext*, bool, clang::DeclContext*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:1095:0
#14 0x00007f8224b8c679 clang::ASTNodeImporter::ImportDefinition(clang::RecordDecl*, clang::RecordDecl*, clang::ASTNodeImporter::ImportDefinitionKind) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:1244:0
#15 0x00007f8224b9ce0e clang::ASTNodeImporter::VisitClassTemplateSpecializationDecl(clang::ClassTemplateSpecializationDecl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:4375:0
#16 0x00007f8224bbed7d clang::declvisitor::Base<clang::declvisitor::make_ptr, clang::ASTNodeImporter, clang::Decl*>::Visit(clang::Decl*) clang/clang1/build/tools/clang/include/clang/AST/DeclNodes.inc:255:0
#17 0x00007f8224bab4a7 clang::ASTImporter::Import(clang::Decl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6835:0
#18 0x00007f8224b89f71 clang::ASTNodeImporter::VisitRecordType(clang::RecordType const*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:771:0
#19 0x00007f8224bbe593 clang::TypeVisitor<clang::ASTNodeImporter, clang::QualType>::Visit(clang::Type const*) clang/clang1/llvm/tools/clang/include/clang/AST/TypeNodes.def:91:0
#20 0x00007f8224bab105 clang::ASTImporter::Import(clang::QualType) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6768:0
#21 0x00007f8224b88def clang::ASTNodeImporter::VisitLValueReferenceType(clang::LValueReferenceType const*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:502:0
#22 0x00007f8224bbe39b clang::TypeVisitor<clang::ASTNodeImporter, clang::QualType>::Visit(clang::Type const*) clang/clang1/llvm/tools/clang/include/clang/AST/TypeNodes.def:67:0
#23 0x00007f8224bab105 clang::ASTImporter::Import(clang::QualType) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6768:0
#24 0x00007f8224b89de2 clang::ASTNodeImporter::VisitAutoType(clang::AutoType const*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:738:0
#25 0x00007f8224bbe653 clang::TypeVisitor<clang::ASTNodeImporter, clang::QualType>::Visit(clang::Type const*) clang/clang1/llvm/tools/clang/include/clang/AST/TypeNodes.def:100:0
#26 0x00007f8224bab105 clang::ASTImporter::Import(clang::QualType) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6768:0
#27 0x00007f8224b88def clang::ASTNodeImporter::VisitLValueReferenceType(clang::LValueReferenceType const*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:502:0
#28 0x00007f8224bbe39b clang::TypeVisitor<clang::ASTNodeImporter, clang::QualType>::Visit(clang::Type const*) clang/clang1/llvm/tools/clang/include/clang/AST/TypeNodes.def:67:0
#29 0x00007f8224bab105 clang::ASTImporter::Import(clang::QualType) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6768:0
#30 0x00007f8224b95fe7 clang::ASTNodeImporter::VisitVarDecl(clang::VarDecl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:2988:0
#31 0x00007f8224bbefbd clang::declvisitor::Base<clang::declvisitor::make_ptr, clang::ASTNodeImporter, clang::Decl*>::Visit(clang::Decl*) clang/clang1/build/tools/clang/include/clang/AST/DeclNodes.inc:441:0
#32 0x00007f8224bab4a7 clang::ASTImporter::Import(clang::Decl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6835:0
#33 0x00007f8224b9e67a operator() clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:4709:0
#34 0x00007f8224b9e67a clang::ASTNodeImporter::ImportDeclGroup(clang::DeclGroupRef)::{lambda(clang::Decl*)#1}::operator()(clang::Decl*) const (clang/clang1/build/bin/../lib/../lib/libclangAST.so.5+0x2df67a)
#35 0x00007f8224bae1fa clang::Decl** std::transform<clang::Decl**, clang::Decl**, clang::ASTNodeImporter::ImportDeclGroup(clang::DeclGroupRef)::{lambda(clang::Decl*)#1}>(clang::Decl**, clang::Decl**, clang::ASTNodeImporter::ImportDeclGroup(clang::DeclGroupRef)::{lambda(clang::Decl*)#1}, clang::ASTNodeImporter::ImportDeclGroup(clang::DeclGroupRef)::{lambda(clang::Decl*)#1}) /usr/include/c++/4.8/bits/stl_algo.h:4926:0
#36 0x00007f8224b9e763 clang::ASTNodeImporter::ImportDeclGroup(clang::DeclGroupRef) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:4712:0
#37 0x00007f8224b9efc4 clang::ASTNodeImporter::VisitDeclStmt(clang::DeclStmt*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:4797:0
#38 0x00007f8224bbfed1 clang::StmtVisitorBase<clang::make_ptr, clang::ASTNodeImporter, clang::Stmt*>::Visit(clang::Stmt*) clang/clang1/build/tools/clang/include/clang/AST/StmtNodes.inc:103:0
#39 0x00007f8224babb30 clang::ASTImporter::Import(clang::Stmt*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6944:0
#40 0x00007f8224ba0d99 clang::ASTNodeImporter::VisitCXXForRangeStmt(clang::CXXForRangeStmt*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:5105:0
#41 0x00007f8224bbfe29 clang::StmtVisitorBase<clang::make_ptr, clang::ASTNodeImporter, clang::Stmt*>::Visit(clang::Stmt*) clang/clang1/build/tools/clang/include/clang/AST/StmtNodes.inc:61:0
#42 0x00007f8224babb30 clang::ASTImporter::Import(clang::Stmt*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6944:0
#43 0x00007f8224bc5194 bool clang::ASTNodeImporter::ImportArrayChecked<clang::Stmt**, clang::Stmt**>(clang::Stmt**, clang::Stmt**, clang::Stmt**)::{lambda(clang::Stmt*)#1}::operator()(clang::Stmt*) const clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:346:0
#44 0x00007f8224bc5194 bool clang::ASTNodeImporter::ImportArrayChecked<clang::Stmt**, clang::Stmt**>(clang::Stmt**, clang::Stmt**, clang::Stmt**)::{lambda(clang::Stmt*)#1}::operator()(clang::Stmt*) const (clang/clang1/build/bin/../lib/../lib/libclangAST.so.5+0x306194)
#45 0x00007f8224bcb9be clang::Stmt** std::transform<clang::Stmt**, clang::Stmt**, bool clang::ASTNodeImporter::ImportArrayChecked<clang::Stmt**, clang::Stmt**>(clang::Stmt**, clang::Stmt**, clang::Stmt**)::{lambda(clang::Stmt*)#1}>(clang::Stmt**, clang::Stmt**, clang::Stmt**, bool clang::ASTNodeImporter::ImportArrayChecked<clang::Stmt**, clang::Stmt**>(clang::Stmt**, clang::Stmt**, clang::Stmt**)::{lambda(clang::Stmt*)#1}) /usr/include/c++/4.8/bits/stl_algo.h:4926:0
#46 0x00007f8224bc5219 bool clang::ASTNodeImporter::ImportArrayChecked<clang::Stmt**, clang::Stmt**>(clang::Stmt**, clang::Stmt**, clang::Stmt**) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:351:0
#47 0x00007f8224bbc67f bool clang::ASTNodeImporter::ImportContainerChecked<llvm::iterator_range<clang::Stmt**>, llvm::SmallVector<clang::Stmt*, 8u> >(llvm::iterator_range<clang::Stmt**> const&, llvm::SmallVector<clang::Stmt*, 8u>&) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:359:0
#48 0x00007f8224b9f1e4 clang::ASTNodeImporter::VisitCompoundStmt(clang::CompoundStmt*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:4816:0
#49 0x00007f8224bbfe71 clang::StmtVisitorBase<clang::make_ptr, clang::ASTNodeImporter, clang::Stmt*>::Visit(clang::Stmt*) clang/clang1/build/tools/clang/include/clang/AST/StmtNodes.inc:79:0
#50 0x00007f8224babb30 clang::ASTImporter::Import(clang::Stmt*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6944:0
#51 0x00007f8224b93842 clang::ASTNodeImporter::VisitFunctionDecl(clang::FunctionDecl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:2570:0
#52 0x00007f8224b93979 clang::ASTNodeImporter::VisitCXXMethodDecl(clang::CXXMethodDecl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:2597:0
#53 0x00007f8224bbef2d clang::declvisitor::Base<clang::declvisitor::make_ptr, clang::ASTNodeImporter, clang::Decl*>::Visit(clang::Decl*) clang/clang1/build/tools/clang/include/clang/AST/DeclNodes.inc:399:0
#54 0x00007f8224bab4a7 clang::ASTImporter::Import(clang::Decl*) clang/clang1/llvm/tools/clang/lib/AST/ASTImporter.cpp:6835:0
#55 0x00007f821d7e3f78 clang::tooling::CrossTranslationUnit::getCrossTUDefinition(clang::FunctionDecl const*, llvm::StringRef, llvm::StringRef, llvm::StringRef, bool) clang/clang1/llvm/tools/clang/lib/Tooling/CrossTranslationUnit.cpp:264:0
#56 0x00007f821dc61fcf clang::ento::AnyFunctionCall::getRuntimeDefinition() const clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/CallEvent.cpp:387:0
#57 0x00007f821dc62c1a clang::ento::CXXInstanceCall::getRuntimeDefinition() const clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/CallEvent.cpp:523:0
#58 0x00007f821dc6317c clang::ento::CXXMemberCall::getRuntimeDefinition() const clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/CallEvent.cpp:623:0
#59 0x00007f821dce3ece clang::ento::ExprEngine::defaultEvalCall(clang::ento::NodeBuilder&, clang::ento::ExplodedNode*, clang::ento::CallEvent const&) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:915:0
#60 0x00007f821dc76ccd clang::ento::CheckerManager::runCheckersForEvalCall(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNodeSet const&, clang::ento::CallEvent const&, clang::ento::ExprEngine&) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/CheckerManager.cpp:619:0
#61 0x00007f821dce2a4e clang::ento::ExprEngine::evalCall(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNode*, clang::ento::CallEvent const&) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:521:0
#62 0x00007f821dce28eb clang::ento::ExprEngine::VisitCallExpr(clang::CallExpr const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngineCallAndReturn.cpp:487:0
#63 0x00007f821dcad0cc clang::ento::ExprEngine::Visit(clang::Stmt const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:1233:0
#64 0x00007f821dca9cb4 clang::ento::ExprEngine::ProcessStmt(clang::CFGStmt, clang::ento::ExplodedNode*) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:548:0
#65 0x00007f821dca8fae clang::ento::ExprEngine::processCFGElement(clang::CFGElement, clang::ento::ExplodedNode*, unsigned int, clang::ento::NodeBuilderContext*) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:394:0
#66 0x00007f821dc92183 clang::ento::CoreEngine::HandlePostStmt(clang::CFGBlock const*, unsigned int, clang::ento::ExplodedNode*) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:532:0
#67 0x00007f821dc90ff4 clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:279:0
#68 0x00007f821dc90bd4 clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:216:0
#69 0x00007f821f6e57ce clang::ento::ExprEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int) clang/clang1/llvm/tools/clang/include/clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h:114:0
#70 0x00007f821f68b4e7 (anonymous namespace)::AnalysisConsumer::ActionExprEngine(clang::Decl*, bool, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:730:0
#71 0x00007f821f68b606 (anonymous namespace)::AnalysisConsumer::RunPathSensitiveChecks(clang::Decl*, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:747:0
#72 0x00007f821f68b2be (anonymous namespace)::AnalysisConsumer::HandleCode(clang::Decl*, unsigned int, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:694:0
#73 0x00007f821f68a2b6 (anonymous namespace)::AnalysisConsumer::HandleDeclsCallGraph(unsigned int) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:510:0
#74 0x00007f821f68a692 (anonymous namespace)::AnalysisConsumer::HandleTranslationUnit(clang::ASTContext&) clang/clang1/llvm/tools/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:561:0
#75 0x00007f8229c0a4da clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) clang/clang1/llvm/tools/clang/lib/Frontend/MultiplexConsumer.cpp:304:0
#76 0x00007f8222ab0ffb clang::ParseAST(clang::Sema&, bool, bool) clang/clang1/llvm/tools/clang/lib/Parse/ParseAST.cpp:161:0
#77 0x00007f8229bde866 clang::ASTFrontendAction::ExecuteAction() clang/clang1/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:1005:0
#78 0x00007f8229bde2d7 clang::FrontendAction::Execute() clang/clang1/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:906:0
#79 0x00007f8229b6e361 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) clang/clang1/llvm/tools/clang/lib/Frontend/CompilerInstance.cpp:981:0
#80 0x00007f822966ad32 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) clang/clang1/llvm/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:251:0
#81 0x000000000044a4b8 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) clang/clang1/llvm/tools/clang/tools/driver/cc1_main.cpp:221:0
#82 0x0000000000440254 ExecuteCC1Tool(llvm::ArrayRef<char const*>, llvm::StringRef) clang/clang1/llvm/tools/clang/tools/driver/driver.cpp:306:0
#83 0x0000000000440e52 main clang/clang1/llvm/tools/clang/tools/driver/driver.cpp:387:0
#84 0x00007f8228f3df45 __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:321:0
#85 0x000000000043dae9 _start (clang/clang1/build/bin/clang-5.0+0x43dae9)
```

**Comment by martong** [comment]
I believe we should not call `setObjectOfFriendDecl` at all, since the `IdentifierNamespace` member of Decl is already imported and copied properly in `Import(Decl*)`.

**Comment by baloghadamsoftware** [comment]
If we do not call it, it does not crash. Is this the correct solution? Do we have a minimal example that can be used as a test?

**Comment by balazske** [comment]
The `IdentifierNamespace` in the imported declaration contains the `IDS_NonMemberOperator` flag, `setObjectOfFriendDecl` does not allow this. According to the documentation:
```
    /// This declaration is a C++ operator declared in a non-class
    /// context.  All such operators are also in IDNS_Ordinary.
    /// C++ lexical operator lookup looks for these.
    IDNS_NonMemberOperator   = 0x0400,
```
This flag should appear in non-class context only (?). But the actual place is inside a class definition (a friend `operator ==` in `std::map`). Probably the documentation is not accurate, or the flag is set to incorrect value?

**Comment by dkrupp** [comment]
create a unit test for testing different order of declarations.

**Comment by balazske** [comment]
The assertion should be solved with #340.
